### ISSUE TITLE
Hashed and salted storage of token_auth

### ISSUE THREAD

**Issue by Joey3000** [matomo-org/matomo#9457]
Authentication tokens have a similar level of confidentiality to passwords. Actually, currently, due to the coupling of the `token_auth` to passwords and their equal power in Piwik, its `token_auth` have exactly the same level of confidentiality as passwords. Yet they are stored in plain text.

The `token_auth` should be stored with the same rigorous hashing and salting as passwords. For the same reasons.

Notes:
- Similar report for strong hashing of passwords: https://github.com/piwik/piwik/issues/5728. Similarly, `token_auth` handling should also use PHP's `password_hash()` for hashing (it's a strong hash and handles salt automatically) and `password_verify()` for value verification (it's safe against timing attacks - see https://github.com/piwik/piwik/issues/9456). They require PHP 5 >= 5.5.0, which should be fine for Piwik 3.0 (see https://github.com/piwik/piwik/issues/8156). Otherwise, there is this: https://github.com/ircmaxell/password_compat.
- This might need an API change, requiring a user name ("login") to always be sent together with `token_auth`. (Otherwise, a "brute-forcing" of the user who the `token_auth` belongs to would be needed. I.e.: hashing of the received `token_auth` with the salt used for each of the users and checking if the result corresponds to any of them. But a DB query for `token_auth`, no matter if hashed or not, could be unsafe against timing attacks - see https://github.com/piwik/piwik/issues/9456. The authentication query should be for the user name ("login"), never for `token_auth`. Just like with passwords.)
  
  This point refers to the fact that currently, in authentication cases without a provided user name (I haven't checked when this happens), the DB is queried using the (plain-text) `token_auth` in https://github.com/piwik/piwik/blob/2.16.0-b1/plugins/Login/Auth.php#L72, which calls https://github.com/piwik/piwik/blob/2.16.0-b1/plugins/UsersManager/Model.php#L173.

**Comment by Joey3000** [comment]
> * This might need an API change, requiring a user name ("login") to always be sent together with `token_auth`. (Otherwise, a "brute-forcing" of the user who the `token_auth` belongs to would be needed. I.e.: hashing of the received `token_auth` with the salt used for each of the users and checking if the result corresponds to any of them. But a DB query for `token_auth`, no matter if hashed or not, could be unsafe against timing attacks - see https://github.com/piwik/piwik/issues/9456. The authentication query should be for the user name ("login"), never for `token_auth`. Just like with passwords.)

>   This point refers to the fact that currently, in authentication cases without a provided user name (I haven't checked when this happens), the DB is queried using the (plain-text) `token_auth` in https://github.com/piwik/piwik/blob/2.16.0-b1/plugins/Login/Auth.php#L72, which calls https://github.com/piwik/piwik/blob/2.16.0-b1/plugins/UsersManager/Model.php#L173.

Another option, without a need for an API change, could be using a part of the transmitted `token_auth` as a "token ID", stored in plain text in a separate DB table column and used for the table look-up. I.e.:
* Using the received "token ID", look up the corresponding hashed `token_auth`
* Using `password_verify()`, securely compare the received `today_auth` (the actual one, without the ID part) to the one looked up
* Use the corresponding user name ("login") if verification successful

**Comment by Joey3000** [comment]
Implementation of `token_auth` hashing would require following additional actions:
* On the part of the user: Update of the transmitted `token_auth` after Piwik has been updated for this. (See next point.)

* On the part of Piwik:
  * When displaying the `token_auth` (in the UsersManager plugin, etc.), show it including the "token ID" in the same string. (Reusing a part of the previously sent `token_auth` will not be possible, due to probable clashes in the ID part. Which needs to be unique among all tokens.)

  * For performance reasons (because password/token hashing with `bcrypt` used internally by the PHP's `password_*()` functions takes time and resources, which is on purpose - to make bruteforcing of leaked hashes slower; as opposed to MD/SHA hash functions, which are used for different purposes and need to be fast): Instead of the requirement to check `token_auth` (see https://developer.piwik.org/guides/security-in-piwik#check-for-the-token_auth), the check would need to be for a different, a **temporary plain text token**.

    Namely: The `token_auth` would be used for authentication on initial request only (before an active session exists), and the temporary plain text token used afterwards. (Handling of the temporary token in plain text is not a problem due to its limited validity period, similarly to a PHP session ID. Unlike `token_auth`, which has a long validity period and in that is similar to a password.) Or should that temporary token be the PHP session ID itself - to automate its expiration, etc.?

  * Because `token_auth` is no longer stored in plain text:
    * If the performance impact is acceptable, and as long as `token_auth` has not been de-coupled from the user password (see https://github.com/piwik/piwik/issues/5703), the displayed (in the UsersManager, etc.) `token_auth` would need to be re-calculated from user password on each display.
    * Otherwise: As is currently the case with, e.g., Github tokens (and Github warns one about that on token creation), a `token_auth` can be shown to the user only once, and never again - namely at the time of its creation. Because it cannot be retrieved from its hash. If a `token_auth` gets lost/forgotten, then a new one needs to be created. (Well, currently it would be re-created from the user password (in which case it will be equal to the old one). But once `token_auth` has been de-coupled from the password, it will be a completely new one.)

I'm not an expert in any way, so I may have mixed things up.

**Comment by tsteur** [comment]
Closing this as a duplicate of https://github.com/matomo-org/matomo/issues/9457
